File: app/src/main/java/com/task/ui/component/news/NewsAdapter.java
package com.task.ui.component.news;

import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.task.R;
import com.task.data.remote.dto.NewsItem;
import com.task.ui.base.listeners.RecyclerItemListener;

import java.util.List;

/**
 * Created by AhmedEltaher on 5/12/2016.
 */

public class NewsAdapter extends RecyclerView.Adapter<NewsViewHolder> {
    private final List<NewsItem> news;
    private RecyclerItemListener onItemClickListener;

    public NewsAdapter(RecyclerItemListener onItemClickListener, List<NewsItem> news) {
        this.onItemClickListener = onItemClickListener;
        this.news = news;
    }

    @Override
    public NewsViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.news_item, parent, false);
        return new NewsViewHolder(view, onItemClickListener);
    }

    @Override
    public void onBindViewHolder(NewsViewHolder holder, int position) {
        holder.bind(position, news.get(position), onItemClickListener);
    }

    @Override
    public int getItemCount() {
        return news.size();
    }
}

File: app/src/main/java/com/task/ui/component/news/newsAdapter/NewsAdapter.kt
package com.task.ui.component.news.newsAdapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.task.data.remote.dto.NewsItem
import com.task.databinding.NewsItemBinding
import com.task.ui.base.listeners.RecyclerItemListener
import com.task.ui.component.news.NewsListViewModel

/**
 * Created by AhmedEltaher
 */

class NewsAdapter(private val newsListViewModel: NewsListViewModel, private val news: List<NewsItem>) : RecyclerView.Adapter<NewsViewHolder>() {

    private val onItemClickListener: RecyclerItemListener = object : RecyclerItemListener {
        override fun onItemSelected(newsItem: NewsItem) {
            newsListViewModel.openNewsDetails(newsItem)
        }

    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): NewsViewHolder {
        val itemBinding = NewsItemBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return NewsViewHolder(itemBinding)
    }

    override fun onBindViewHolder(holder: NewsViewHolder, position: Int) {
        holder.bind(news[position], onItemClickListener)
    }

    override fun getItemCount(): Int {
        return news.size
    }
}

